core/extension/gdextension_interface.h

```diff
@@ -343,6 +343,47 @@ typedef struct {
 	GDExtensionVariantPtr *default_arguments;
 } GDExtensionClassMethodInfo;
 
+typedef void (*GDExtensionCallableCustomCall)(void *callable_userdata, const GDExtensionConstVariantPtr *p_args, GDExtensionInt p_argument_count, GDExtensionVariantPtr r_return, GDExtensionCallError *r_error);
+typedef GDExtensionBool (*GDExtensionCallableCustomIsValid)(void *callable_userdata);
+typedef void (*GDExtensionCallableCustomFree)(void *callable_userdata);
+
+typedef uint32_t (*GDExtensionCallableCustomHash)(void *callable_userdata);
+typedef GDExtensionBool (*GDExtensionCallableCustomEqual)(void *callable_userdata_a, void *callable_userdata_b);
+typedef GDExtensionBool (*GDExtensionCallableCustomLessThan)(void *callable_userdata_a, void *callable_userdata_b);
+
+typedef void (*GDExtensionCallableCustomToString)(void *callable_userdata, GDExtensionBool *r_is_valid, GDExtensionStringPtr r_out);
+
+typedef struct {
+	/* Only `call_func` is strictly required, however, `object` should be passed if its not a static method.
+	 *
+	 * `hash_func`, `equal_func`, and `less_than_func` are optional. If not provided both `call_func` and
+	 * `callable_userdata` together are used as the identity of the callable for hashing and comparison purposes.
+	 *
+	 * The hash returned by `hash_func` is cached, `hash_func` will not be called more than once per callable.
+	 *
+	 * `is_valid_func` is necessary if the validity of the callable can change before destruction.
+	 *
+	 * `free_func` is necessary if `callable_userdata` needs to be cleaned up when the callable is freed.
+	 *
+	 * `userdata_tag` is used when calling `callable_custom_get_userdata` to identify and ensure that only the
+	 * correct type of userdata is returned.
+	 */
+	void *callable_userdata;
+	void *userdata_tag;
```

Comment:
Why is `userdata_tag` needed? Couldn't you just make your userdata be a struct like:

```cpp
struct CallableUserdata {
  Type userdata_tag;
  UnderlyingData *data;
};
```

... and then it's up to the code in the binding and callbacks to check that the data has the right type?

Reply:
Multiple extensions could have their own custom callables with different userdata layouts, `userdata_tag` ensures they don't access memory from some other extension.

Reply:
Ah, ok, that makes sense!

In that case, let's call this `token` (instead of `userdata_tag`) to match the other situations where we need to group some data by extension. And then we should also document it as required, and a clearer explanation of what it's supposed to represent.

Reply:
Oh, and all the comparisions in `default_compare_equal()`, `default_compare_less()`, etc should check the `token` first, so we "short-circuit" early if we're attempting to compare callables from different extensions.

Reply:
Ok, `token` sounds good, will change.

Short-circuiting on `token` doesn't gain anything from a functional / performance pov (would probably be immeasurably worse actually as `token` will have less variability)

Doing so world probably make more sense semantically tho, to help readability. I guess this is what you meant.

Do you have a suggestion on how to improve the docs?

Reply:
> Multiple extensions could have their own custom callables with different userdata layouts, userdata_tag ensures they don't access memory from some other extension.

Actually, thinking about this a little more, I don't think we need to add this to prevent `equal_func` and `less_than_func` from accessing memory from other extensions. `Callable` already only compares `CallableCustom`s that have the same comparison function. Even if two GDExtensions both use godot-cpp, they will statically link with it, so each extension will provide a pointer to a different comparison function. This is why we don't need to worry about a `CallableCustom` from GDExtension being compared with one created by `callable_mp()` in the engine itself.

I think the only case where we really need to differentiate which extension a `CallableCustom` comes from is in "callable_custom_get_userdata" exposed in `gdextension_interface.h`.

> Do you have a suggestion on how to improve the docs?

Sure, how about this:

```cpp
	/* Only `call_func` and `token` are strictly required, however, `object` should be passed if its not a static method.
	 *
	 * `token` should point to an address that uniquely identifies the GDExtension (for example, the `GDExtensionClassLibraryPtr` passed to the entry symbol function.
	 *
	 * `hash_func`, `equal_func`, and `less_than_func` are optional. If not provided both `call_func` and
	 * `callable_userdata` together are used as the identity of the callable for hashing and comparison purposes.
	 *
	 * The hash returned by `hash_func` is cached, `hash_func` will not be called more than once per callable.
	 *
	 * `is_valid_func` is necessary if the validity of the callable can change before destruction.
	 *
	 * `free_func` is necessary if `callable_userdata` needs to be cleaned up when the callable is freed.
	 */
```

Reply:
I think you're right about not needing `token` in the comparison functions. I'll make the changes a bit later.